### PR TITLE
Poc/umr uct

### PR DESCRIPTION
**Pull request by artpol84**
Work in progress.
Compiles, but not runtime-tested

### CODE REVIEW

**Comment by artpol84** [comment]
I figured that we can get away with only "struct" type. No need for "stride" and "interleave".

File: src/ucp/dt/dt_struct.c [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+
+static void _set_depth(ucp_dt_struct_t *s)
+{
+    size_t i, depth = 0;
+
+    for (i = 0; i < s->desc_count; i++) {
+        ucp_struct_dt_desc_t *dsc = &s->desc[i];
+        switch (dsc->dt & UCP_DATATYPE_CLASS_MASK) {
+        case UCP_DATATYPE_CONTIG:
+            depth = ucs_max(depth, 1);
```
Comment (brminich) [inline src/ucp/dt/dt_struct.c]: do we really need to set 1 here?

File: src/ucp/dt/dt_struct.c [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+    size_t iov_cnt = s->desc_count;
+    uct_md_h md_p;
+    uct_mem_h memh;
+    uct_completion_t comp;
+
+    /* On the leaf level, all datatypes are contig */
+    for (i = 0; i < s->desc_count; i++) {
+        iovs[i].buffer = ptr + s->desc[i].displ;
+        iovs[i].length = ucp_contig_dt_length(s->desc[i].dt, 1);
+        // iovs[i].count = ????
```
Comment (brminich) [inline src/ucp/dt/dt_struct.c:340]: count is not used with our iovs

File: src/ucp/dt/dt_struct.c [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+        iovs[i].stride = s->desc[i].extent;
+        // iovs[i].memh = ??
```
Comment (brminich) [inline src/ucp/dt/dt_struct.c:342]: probably we have to register the whole (contig) struct memory region during struct dt creation and save memh in the struct. Its caching should be handled by rcache

File: src/ucp/dt/dt_struct.c [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+    khiter_t k;
+    k = kh_get(dt_struct, &s->hash, (uint64_t)ptr);
+    return (k != kh_end(&s->hash));
+}
+
+static int _is_primitive_strided(ucp_dt_struct_t *s)
+{
+    size_t i;
+    /* Has only 2 levels of nesting */
+    if (!(s->depth = 2)){
```
Comment (brminich) [inline src/ucp/dt/dt_struct.c]: did you mean `s != 2`
why only 2 ?

File: src/ucp/dt/dt_struct.c [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+        return 0;
+    }
+
+    /* On the leaf level, all datatypes are contig */
+    for (i = 0; i < s->desc_count; i++) {
```
Comment (brminich) [inline src/ucp/dt/dt_struct.c:318]: so, we do not allow nesting striding?
worth to check and print error during dt creation

File: src/ucp/dt/dt_struct.c [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+    uct_completion_t comp;
+
+    /* On the leaf level, all datatypes are contig */
+    for (i = 0; i < s->desc_count; i++) {
+        iovs[i].buffer = ptr + s->desc[i].displ;
+        iovs[i].length = ucp_contig_dt_length(s->desc[i].dt, 1);
+        // iovs[i].count = ????
+        iovs[i].stride = s->desc[i].extent;
+        // iovs[i].memh = ??
+    }
```
Comment (brminich) [inline src/ucp/dt/dt_struct.c:343]: maybe it is worth to define `init|fini` functions to bind dt with exact memory and assign mem handlers, like we have for generic types (start_pack|upack)

File: src/ucp/dt/dt_struct.c [hunk 7]
```diff
@@ -0,0 +1,365 @@
...
+
+    /* First, find the sequential number of the repetition that holds this
+     * offset
+     */
+    *rep_num = offset / s->step_len;
+    if( !(*rep_num < s->rep_count) ) {
+        /* Shouldn't happen */
+        return -1;
+    }
+    toffs = (*rep_num) * s->step_len;
```
Comment (brminich) [inline src/ucp/dt/dt_struct.c:89]: it should be equal to offset, right?

File: src/ucp/dt/dt_struct.c [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+    *datatype_p = ((uintptr_t)dt) | UCP_DATATYPE_STRUCT;
+    return UCS_OK;
+}
+
+void ucp_dt_destroy_struct(ucp_datatype_t datatype_p)
+{
+    ucp_dt_struct_t *dt;
+    dt = ucp_dt_struct(datatype_p);
+    ucs_free(dt->desc);
+    ucs_free(dt);
```
Comment (brminich) [inline src/ucp/dt/dt_struct.c:257]: would need to clear cache here (amd unregister umrs)

File: src/ucp/dt/dt_struct.c [hunk 9]
```diff
@@ -0,0 +1,365 @@
...
+    if (ret != 0) {
+        ucs_free(dt);
+        return UCS_ERR_NO_MEMORY;
+    }
+    memcpy(dt->desc, desc_ptr, sizeof(*desc_ptr) * desc_count);
+    dt->desc_count = desc_count;
+    dt->rep_count = rep_count;
+    _set_length(dt);
+    _set_depth(dt);
+    *datatype_p = ((uintptr_t)dt) | UCP_DATATYPE_STRUCT;
```
Comment (brminich) [inline src/ucp/dt/dt_struct.c:248]: would need to init hash here (if more than 1 element, or single but non-contig element)

File: src/ucp/dt/dt_struct.c (hunk 4 above)
Comment (artpol84) [reply src/ucp/dt/dt_struct.c]:
```suggestion
    if (s->depth != 2){
```

File: src/ucp/dt/dt_struct.c [hunk 10]
```diff
@@ -0,0 +1,365 @@
...
+    /* wait for completion */
+
+    _to_cache(s, ptr, memh);
+    return 1;
+}
+
+int ucp_dt_struct_register(uct_ep_h ep, ucp_datatype_t dt, void *ptr)
+{
+    ucp_dt_struct_t *s = ucp_dt_struct(dt);
+
```
Comment (artpol84) [inline src/ucp/dt/dt_struct.c:356]: Calculate extent and register contig region as the base mem handler

File: src/ucp/tag/tag_send.c [hunk 11]
```diff
@@ -38,6 +38,8 @@ ucp_tag_get_rndv_threshold(const ucp_request_t *req, size_t count,
         return ucs_min(rndv_rma_thresh, rndv_am_thresh);
     case UCP_DATATYPE_GENERIC:
         return rndv_am_thresh;
+    case UCP_DATATYPE_STRUCT:
+        /* TODO: ???? */
```
Comment (artpol84) [inline src/ucp/tag/tag_send.c]: (size_t)(-1)
```suggestion
        return (size_t)(-1);
```

File: src/ucp/dt/dt_struct.c (hunk 1 above)
Comment (artpol84) [reply src/ucp/dt/dt_struct.c]:
```suggestion
            depth = 0;
```

**Pull request merged**